### PR TITLE
Exactly-once for JMS source

### PR DESCRIPTION
**Pull request by viliam-durina**

### CODE REVIEW

**Comment by viliam-durina** [comment]
Summary of breaking changes:
- Removed `sessionFn` and `flushFn` from the `JmsSourceBuilder`. I doubt anybody used them because non-default functions couldn't achieve even at-least-once behavior.
- If guarantee == EXACTLY_ONCE, the `getMessageID` is required to return non-null.

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/JmsSourceBuilder.java [hunk 1]
```diff
@@ -46,114 +48,171 @@
...
+        checkSerializable(consumerFn, "consumerFn");
+        this.consumerFn = consumerFn;
         return this;
     }
 
     /**
-     * Sets the name of the destination. If {@code consumerFn} is provided this
-     * parameter is ignored.
+     * Configures the function to extract IDs from the messages, if
+     * exactly-once guarantee is used. If lower guarantee is used, this
```
Comment (mtopolnik) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/JmsSourceBuilder.java]: If _a_ lower guarantee...

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/JmsSourceBuilder.java [hunk 2]
```diff
@@ -46,114 +48,171 @@
...
 
     /**
-     * Sets the function which creates the message consumer from session.
-     * <p>
-     * If not provided, the builder creates a function which uses {@code
-     * Session#createConsumer(Destination destination)} to create the consumer.
-     * Either {@code consumerFn} or {@code destinationName} should be set. See
-     * {@link #destinationName(String)}.
+     * Sets the maximum processing guarantee for the source. You can use it to
+     * decrease the guarantee of this source compared to the job's guarantee.
```
Comment (mtopolnik) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/JmsSourceBuilder.java]: "decrease" -> "reduce"

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/JmsSourceBuilder.java [hunk 3]
```diff
@@ -46,114 +48,171 @@
...
+     * If you configure stronger guarantee than the job has, the job's
```
Comment (mtopolnik) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/JmsSourceBuilder.java]: _a_ stronger guarantee

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/JmsSourceBuilder.java [hunk 4]
```diff
@@ -46,114 +48,171 @@
...
 
     /**
-     * Sets the function which creates the session from connection.
-     * <p>
-     * If not provided, the builder creates a function which uses {@code
-     * Connection#createSession(boolean transacted, int acknowledgeMode)} to
-     * create the session. See {@link #sessionParams(boolean, int)}.
+     * Sets the function which creates the message consumer from session.
+     *
+     * <p>If not provided, {@code Session#createConsumer(destinationName)} is used
```
Comment (mtopolnik) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/JmsSourceBuilder.java]: I would strongly prefer you not introducing this style of using `<p>`, which is inconsistent with the rest of the codebase. It also makes it harder for me to spot any missing `<p>`s, a problem which becomes apparent only in the generated Javadoc.

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/JmsSourceBuilder.java [hunk 5]
```diff
@@ -46,114 +48,171 @@
...
+     * be used to deduplicate the re-delivered messages.
+     *
+     * <p>If you use a non-durable consumer with a topic, set the max-guarantee
+     * to NONE - the broker will not redeliver the unacknowledged messages
+     * anyway. If you didn't specify your own {@link #consumerFn(FunctionEx)},
+     * Jet will do it for you automatically because the consumer for the topic
+     * is non-durable by default.
+     *
+     * <p>The default is {@link ProcessingGuarantee#EXACTLY_ONCE}, which means
+     * that source's guarantee will match the job's guarantee.
```
Comment (mtopolnik) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/JmsSourceBuilder.java]: _the_ source's guarantee

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/Sources.java [hunk 6]
```diff
@@ -1020,7 +1021,7 @@ public static FileSourceBuilder filesBuilder(@Nonnull String directory) {
      * append the lines. However, it might not work as expected because some
      * editors write to a temp file and then rename it or append extra newline
      * character at the end which gets overwritten if more text is added in the
-     * editor. Best way to append is to use {@code echo text >> yourfile}.
+     * editor. Best way to append is to use {@code echo text >> yourFile}.
```
Comment (mtopolnik) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/Sources.java]: _The_ best way

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/JmsSourceBuilder.java [hunk 7]
```diff
@@ -46,114 +48,171 @@
...
+    @Nonnull
+    public JmsSourceBuilder maxGuarantee(@Nonnull ProcessingGuarantee guarantee) {
+        maxGuarantee = checkNotNull(guarantee);
         return this;
     }
 
     /**
-     * Sets the function which commits the session after consuming each message.
+     * Specifies whether the MessageConsumer of the JMS topic is shared, that
+     * is when {@code createSharedConsumer()} or {@code
```
Comment (mtopolnik) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/JmsSourceBuilder.java]: _when_ -> _whether_

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/JmsSourceBuilder.java [hunk 8]
```diff
@@ -46,114 +48,171 @@
...
+     * createSharedDurableConsumer()} was used to create it in the {@link
+     * #consumerFn(FunctionEx)}.
+     * <p>
+     * If the consumer is not shared, only single processor on single member
```
Comment (mtopolnik) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/JmsSourceBuilder.java]: _a_ single processor on _a_ single member

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/JmsSourceBuilder.java [hunk 9]
```diff
@@ -46,114 +48,171 @@
...
+     * will connect to the broker to receive the messages. If you set this
+     * parameter to {@code true} for a non-shared consumer, all messages will
+     * be emitted on every member, leading to duplicate processing.
      * <p>
-     * If not provided, the builder creates a no-op consumer.
+     * The consumer for a queue is always assumed to be shared, regardless of
```
Comment (mtopolnik) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/JmsSourceBuilder.java]: _A_ consumer _of_ a queue

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/Sources.java [hunk 10]
```diff
@@ -1056,39 +1062,46 @@ public static FileSourceBuilder filesBuilder(@Nonnull String directory) {
...
+     * <p>This source supports exactly-once and at-least-once mode, see {@link
+     * JmsSourceBuilder#maxGuarantee(ProcessingGuarantee)} for more
+     * information.
+     *
+     * <p>IO failures should be handled by the JMS provider. If any JMS
+     * operation throws an exception, the job will fail. Most of the providers
+     * offer a configuration parameter to enable auto-reconnection, refer to
+     * provider documentation for details.
+     *
+     * <p>Default local parallelism for this processor is 1.
```
Comment (mtopolnik) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/Sources.java]: _The default_

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/Sources.java [hunk 11]
```diff
@@ -1104,28 +1117,25 @@ public static JmsSourceBuilder jmsQueueBuilder(SupplierEx<? extends ConnectionFa
...
```
Comment (mtopolnik) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/Sources.java]: _The_ default

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/JmsSourceBuilder.java (hunk 4 above)
Comment (viliam-durina) [reply hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/JmsSourceBuilder.java]: On the other hand it improves source code readability when you're _not_ interested in checking for missing `<p>`. And, most importantly, there's tool support for this style: I use "Wrap to Column" plugin and it doesn't support the other style without manually selecting a block to reformat. I tweaked the text many many times and I hate to reflow manually or being forced to select it before reformatting.

And I already sneaked quite a few instances into the code base :)

Comment (cangencer) [reply hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/JmsSourceBuilder.java]: I think we don't terminate <p> typically, but it should be on a separate line, not part of the paragraph.

**Comment by viliam-durina** [comment]
verify

**Pull request merged**